Rework RhGetRuntimeHelperForType in C#

kant2002 opened the pull request:
as per this discussion https://github.com/dotnet/runtimelab/pull/1294#issuecomment-874430938

kant2002 committed:
Rework RhGetRuntimeHelperForType in C#
This is required to make compile on ARM

jkotas commented:
You may also need to change `GenericLookupResultReferenceType LookupResultReferenceType(NodeFactory factory)` for affected dictionary cell types to return `GenericLookupResultReferenceType.Direct `.

kant2002 committed:
Changes based on PR feedback

kant2002 committed:
Remvoe last Indirect lookups.
Technically there no indirecion marcro left, but interesting, should I keep GenericLookupResultReferenceType as it is. Probably this is some core feature which used occasionallly to get tactical results

kant2002 committed:
Half revert previous attempt in order to guess if this is proper node.

kant2002 committed:
This seems to be best option, and I have to look what's going on

kant2002 committed:
Seems to be it

MichalStrehovsky commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/RuntimeImports.cs:
```diff
@@ -487,9 +487,63 @@ internal enum RuntimeHelperKind
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        [RuntimeImport(RuntimeLibrary, "RhGetRuntimeHelperForType")]
-        internal static extern unsafe IntPtr RhGetRuntimeHelperForType(EETypePtr pEEType, RuntimeHelperKind kind);
+        internal static unsafe IntPtr RhGetRuntimeHelperForType(EETypePtr pEEType, RuntimeHelperKind kind)
```
Instead of inlining this into RuntimeImports.cs, could you put the implementation as a `RuntimeExport` into src\coreclr\nativeaot\Runtime.Base\src\System\Runtime\RuntimeExports.cs? Use `EEType*` instead of `EETypePtr` there.

RuntimeHelperKind enum will have to move to src\coreclr\tools\Common\Internal\Runtime\RuntimeConstants.cs but that's a better spot for it.

Then you can keep this InternalCall as-is.

MichalStrehovsky reviewed (commented):
There is one more area that needs adjustment - the runtime type loader.

Search for `HelperForType` in src\coreclr\nativeaot\System.Private.TypeLoader. The places that try to dereference the result should no longer do that. There should be 4 hits.

kant2002 committed:
Do not perform indrection in the related DictionaryCell
Maintain separation of the runtime implementation

kant2002 committed:
Remove not needed code for this change

kant2002 commented:
Tests do not run for some reason.

MichalStrehovsky commented on src/coreclr/nativeaot/System.Private.TypeLoader/src/Internal/Runtime/TypeLoader/GenericDictionaryCell.cs:
```diff
@@ -294,7 +294,7 @@ internal override IntPtr Create(TypeBuilder builder)
             internal override unsafe IntPtr CreateLazyLookupCell(TypeBuilder builder, out IntPtr auxResult)
             {
                 auxResult = IntPtr.Zero;
-                return *(IntPtr*)Create(builder);
```
This one is still indirected.

MichalStrehovsky commented on src/coreclr/tools/Common/Internal/Runtime/RuntimeConstants.cs:
```diff
@@ -57,4 +57,14 @@ internal static class SpecialDispatchMapSlot
         public const ushort Reabstraction = 0xFFFF;
     }
+
+    // Keep in sync with RH\src\rtu\runtimeinstance.cpp
```
We only have one copy of this now - no longer need to keep in sync with anything.

MichalStrehovsky commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/RuntimeExports.cs:
```diff
@@ -450,5 +450,65 @@ public static int RhGetThunkSize()
+                return (IntPtr)(delegate*<EEType*, object, void>)&TypeCast.CheckVectorElemAddr;
+
+            default:
+                throw new NotImplementedException();
```
Throwing from here won't lead to anything good. Maybe assert and return `IntPtr.Zero`?

MichalStrehovsky reviewed (commented):
Looks good otherwise! Thanks!

MichalStrehovsky commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/RuntimeExports.cs:
```diff
@@ -450,5 +450,65 @@ public static int RhGetThunkSize()
             return InternalCalls.RhpGetThunkSize();
         }
+
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
```
The MethodImpl is not needed.

kant2002 committed:
Address PR feedback

kant2002 committed:
Fix compilation error

MichalStrehovsky reviewed (approved):
Thank you!
